### ISSUE TITLE
Handling circular object in stringify

### ISSUE THREAD

**Issue by vojtatranta** [dominictarr/JSONStream#102]
We have an uncaught exception everytime we try to stringify circular object.

```js
state = {}
state['circular'] = state

rs = new Readable({ objectMode: true })
rs.push(state)
rs.push(null)
jsonStream = rs.pipe(JSONStream.stringify(false))

writeStream = fs.createWriteStream(filename)
jsonStream.pipe(writeStream)

// this catches no error
 writeStream.on('error', err => {
   console.error(
     'StateManager#saveStateAsync()' +
     ' -> state file cannot be saved',
     err
   )
})
```
This throws an uncaught exception no matter where I put `on('error')` listener.

```
/Users/vojtatranta/source/avocode/avocode-next/script/test.gulp.coffee:69
      throw err;
      ^
TypeError: Converting circular structure to JSON
  at Object.stringify (native)
  at Stream.<anonymous> (/Users/vojtatranta/source/avocode/avocode-next/node_modules/JSONStream/index.js:159:23)
  at Stream.stream.write (/Users/vojtatranta/source/avocode/avocode-next/node_modules/JSONStream/node_modules/through/index.js:26:11)
  ....
  at process._tickDomainCallback (node.js:391:17)
```

Basically, I think that there is an error in JSONStream source code here: https://github.com/dominictarr/JSONStream/blob/master/index.js#L158

Where is used `JSON.stringify` without `try-catch` block around it. Therefore, no `stream.emit('error')` is called resulting in this error which cannot be caught, because it is a result of `at process._tickDomainCallback (node.js:391:17)`.

What do you guys think?

**Comment by dominictarr** [comment]
you are correct. unfortunately, you cannot stringify a circular object. on the other hand, that keeps JSON simple. it's best to avoid circular objects. closed by #103

**Comment by vojtatranta** [comment]
@dominictarr yup yup, you should not have such structure. But it might happen and you want to be able to catch such error not to show it to poor user.
